File: bindings/go/batch.go
package blazen

import (
	"context"

	uniffiblazen "github.com/zachhandley/Blazen/bindings/go/internal/uniffi/blazen"
)

// BatchItem is the per-request outcome inside a [BatchResult]'s Responses
// slice. It is a sealed sum type: only [BatchItemSuccess] and
// [BatchItemFailure] satisfy it, matching the FFI's two-variant enum.
//
// Type-switch on the concrete variant to consume an outcome:
//
//	switch v := item.(type) {
//	case blazen.BatchItemSuccess:
//	    use(v.Response)
//	case blazen.BatchItemFailure:
//	    log(v.ErrorMessage)
//	}
type BatchItem interface {
	batchItem()
}

// BatchItemSuccess is the [BatchItem] variant carrying a completed
// [CompletionResponse]. Slot ordering in [BatchResult.Responses] matches
// the input request slice passed to [CompleteBatch].
type BatchItemSuccess struct {
	Response CompletionResponse
}

func (BatchItemSuccess) batchItem() {}

// BatchItemFailure is the [BatchItem] variant carrying a per-request
// error. ErrorMessage mirrors the Display form of the underlying
// BlazenError on the Rust side; the structured variant is flattened to a
// string at the FFI boundary because nested-enum errors do not survive
// every UniFFI target language. Callers wanting typed errors should
// dispatch requests individually via [CompletionModel.Complete].
type BatchItemFailure struct {
	ErrorMessage string
}

func (BatchItemFailure) batchItem() {}

// BatchResult is the aggregated outcome of a [CompleteBatch] /
// [CompleteBatchBlocking] call.
//
// Responses is parallel to the input requests slice: Responses[i]
// corresponds to requests[i]. Per-request errors appear as
// [BatchItemFailure] entries and do NOT cause [CompleteBatch] itself to
// return a Go-level error — only batch-level validation failures do
// (e.g. a malformed ParametersJSON / ResponseFormatJSON that fails to
// reach the provider at all).
//
// TotalUsage and TotalCostUSD aggregate only the successful responses;
// failed slots contribute zero. A TotalCostUSD of 0.0 is ambiguous
// between "no cost incurred" and "no provider reported cost data" — the
// wire format does not distinguish the two.
type BatchResult struct {
	Responses    []BatchItem
	TotalUsage   TokenUsage
	TotalCostUSD float64
}

// batchItemFromFFI converts the generated FFI BatchItem sum type into
// the wrapper's sealed interface. Unknown variants are downgraded to a
// [BatchItemFailure] with an explanatory message — this is defensive and
// should never trip in practice unless the generated bindings drift
// ahead of this wrapper.
func batchItemFromFFI(item uniffiblazen.BatchItem) BatchItem {
	switch v := item.(type) {
	case uniffiblazen.BatchItemSuccess:
		return BatchItemSuccess{Response: completionResponseFromFFI(v.Response)}
	case uniffiblazen.BatchItemFailure:
		return BatchItemFailure{ErrorMessage: v.ErrorMessage}
	default:
		return BatchItemFailure{ErrorMessage: "unknown batch item variant"}
	}
}

// batchResultFromFFI converts the generated FFI BatchResult into the
// wrapper form, translating each per-request slot and renaming the
// generated TotalCostUsd field to the idiomatic Go TotalCostUSD.
func batchResultFromFFI(r uniffiblazen.BatchResult) *BatchResult {
	items := make([]BatchItem, len(r.Responses))
	for i, ffiItem := range r.Responses {
		items[i] = batchItemFromFFI(ffiItem)
	}
	return &BatchResult{
		Responses:    items,
		TotalUsage:   tokenUsageFromFFI(r.TotalUsage),
		TotalCostUSD: r.TotalCostUsd,
	}
}

// CompleteBatch dispatches a batch of completion requests through the
// supplied model with a bounded concurrency cap.
//
// maxConcurrency = 0 means unlimited (every request dispatched in
// parallel, bounded only by the provider's own rate limits). For most
// callers, 4-8 is a reasonable starting point.
//
// Honours ctx cancellation by detaching from the wait: if ctx fires
// before the batch completes, the function returns ctx.Err() and the
// Rust-side batch continues to completion in the background.
// Cancellation propagation into the runtime is a known gap pending an
// upstream UniFFI feature.
//
// A non-nil error return indicates a batch-level failure (e.g. the
// model handle is closed, or every request failed input validation
// before dispatch). Per-request failures surface as [BatchItemFailure]
// entries inside the returned [BatchResult] and do not produce an
// error here.
func CompleteBatch(ctx context.Context, model *CompletionModel, requests []CompletionRequest, maxConcurrency uint32) (*BatchResult, error) {
	ensureInit()
	if model == nil || model.inner == nil {
		return nil, &ValidationError{Message: "model is nil or closed"}
	}
	ffiReqs := make([]uniffiblazen.CompletionRequest, len(requests))
	for i, r := range requests {
		ffiReqs[i] = r.toFFI()
	}
	type batchOutcome struct {
		res uniffiblazen.BatchResult
		err error
	}
	done := make(chan batchOutcome, 1)
	go func() {
		res, err := uniffiblazen.CompleteBatch(model.inner, ffiReqs, maxConcurrency)
		done <- batchOutcome{res: res, err: err}
	}()
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case r := <-done:
		if r.err != nil {
			return nil, wrapErr(r.err)
		}
		return batchResultFromFFI(r.res), nil
	}
}

// CompleteBatchBlocking is the synchronous form of [CompleteBatch]. It
// does not accept a [context.Context] and blocks the calling goroutine
// until the entire batch completes. Prefer [CompleteBatch] in
// long-running services where cancellation matters; use this in short
// scripts or main() functions where the async wiring is overkill.
//
// The semantics of maxConcurrency, the [BatchResult] return value, and
// the distinction between batch-level errors (returned here) versus
// per-request failures (carried inside [BatchResult.Responses]) are
// identical to [CompleteBatch].
func CompleteBatchBlocking(model *CompletionModel, requests []CompletionRequest, maxConcurrency uint32) (*BatchResult, error) {
	ensureInit()
	if model == nil || model.inner == nil {
		return nil, &ValidationError{Message: "model is nil or closed"}
	}
	ffiReqs := make([]uniffiblazen.CompletionRequest, len(requests))
	for i, r := range requests {
		ffiReqs[i] = r.toFFI()
	}
	res, err := uniffiblazen.CompleteBatchBlocking(model.inner, ffiReqs, maxConcurrency)
	if err != nil {
		return nil, wrapErr(err)
	}
	return batchResultFromFFI(res), nil
}

File: crates/blazen-uniffi/src/batch.rs
//! Batch completion surface for the UniFFI bindings.
//!
//! Thin wrapper around upstream [`blazen_llm::batch::complete_batch`], which
//! runs N [`CompletionRequest`]s through a single [`CompletionModel`] with
//! bounded concurrency and aggregates usage / cost across the successful
//! responses.
//!
//! ## Result ordering and partial failure
//!
//! Responses are returned in the same order as the input requests. Individual
//! requests may fail without short-circuiting the rest of the batch — each
//! slot in [`BatchResult::responses`] is a [`BatchItem::Success`] or
//! [`BatchItem::Failure`] independent of its siblings. This matches the
//! upstream contract (`Vec<Result<CompletionResponse, _>>` with per-slot
//! errors).
//!
//! ## Reused upstream API
//!
//! No reimplementation of the fan-out happens here; the parallelism, the
//! concurrency semaphore, and the usage / cost accumulators all live in
//! upstream `complete_batch`.

use std::sync::Arc;

use blazen_llm::batch::{
    BatchConfig as CoreBatchConfig, BatchResult as CoreBatchResult,
    complete_batch as core_complete_batch,
};
use blazen_llm::types::CompletionRequest as CoreCompletionRequest;

use crate::errors::{BlazenError, BlazenResult};
use crate::llm::{CompletionModel, CompletionRequest, CompletionResponse, TokenUsage};
use crate::runtime::runtime;

// ---------------------------------------------------------------------------
// Wire-format records
// ---------------------------------------------------------------------------

/// Per-request outcome within a [`BatchResult`].
///
/// Slot `i` of [`BatchResult::responses`] corresponds to input request `i`.
/// Successful slots carry the [`CompletionResponse`]; failed slots carry an
/// `error_message` only (the structured `BlazenError` variant doesn't survive
/// nesting inside a `uniffi::Enum` cleanly across all four target languages,
/// so the message is flattened to a string here — foreign callers wanting
/// typed errors should run requests individually).
#[derive(Debug, Clone, uniffi::Enum)]
pub enum BatchItem {
    /// The request completed and the model returned a response.
    Success { response: CompletionResponse },
    /// The request failed. The message mirrors the `Display` form of the
    /// underlying [`BlazenError`].
    Failure { error_message: String },
}

/// Outcome of a [`complete_batch`] call.
///
/// `total_usage` and `total_cost_usd` aggregate only the successful responses
/// — failed slots contribute zero. When no provider reports cost data the
/// total is `0.0` (the wire format does not distinguish "zero" from "unknown").
#[derive(Debug, Clone, uniffi::Record)]
pub struct BatchResult {
    /// One slot per input request, in the same order.
    pub responses: Vec<BatchItem>,
    /// Aggregated token usage across successful responses.
    pub total_usage: TokenUsage,
    /// Aggregated USD cost across successful responses.
    pub total_cost_usd: f64,
}

// ---------------------------------------------------------------------------
// Public surface
// ---------------------------------------------------------------------------

/// Run a batch of completion requests with bounded concurrency.
///
/// - `model`: the model to drive (one provider, one model id; for cross-model
///   batches dispatch from foreign code instead).
/// - `requests`: the requests to send, in order. Each is converted to the
///   upstream wire format before dispatch; conversion errors short-circuit
///   the entire batch (the request list is rejected as a whole, since a bad
///   schema means the batch was misconfigured).
/// - `max_concurrency`: hard cap on in-flight requests. `0` means unlimited
///   (all dispatched in parallel).
///
/// Returns a [`BatchResult`] with per-request outcomes and aggregated
/// usage / cost. Individual request failures appear in
/// [`BatchResult::responses`] as [`BatchItem::Failure`] — they do not cause
/// this function itself to return `Err`.
///
/// # Errors
///
/// Returns [`BlazenError::Validation`] if any input request fails to convert
/// to the upstream wire format (typically a malformed `parameters_json` or
/// `response_format_json` payload).
#[uniffi::export(async_runtime = "tokio")]
pub async fn complete_batch(
    model: Arc<CompletionModel>,
    requests: Vec<CompletionRequest>,
    max_concurrency: u32,
) -> BlazenResult<BatchResult> {
    let core_requests = build_core_requests(requests)?;
    let config = CoreBatchConfig::new(max_concurrency as usize);
    let core_result = core_complete_batch(model.inner.as_ref(), core_requests, config).await;
    Ok(batch_result_from_core(core_result))
}

/// Synchronous variant of [`complete_batch`] — blocks the current thread on
/// the shared Tokio runtime.
///
/// # Errors
///
/// Same as [`complete_batch`].
#[uniffi::export]
pub fn complete_batch_blocking(
    model: Arc<CompletionModel>,
    requests: Vec<CompletionRequest>,
    max_concurrency: u32,
) -> BlazenResult<BatchResult> {
    runtime().block_on(async move { complete_batch(model, requests, max_concurrency).await })
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Convert wire [`CompletionRequest`]s into upstream [`CoreCompletionRequest`]s.
///
/// Errors out as soon as any request fails to convert — a malformed request
/// in the middle of a batch usually means the entire batch is misconfigured.
fn build_core_requests(
    requests: Vec<CompletionRequest>,
) -> BlazenResult<Vec<CoreCompletionRequest>> {
    requests
        .into_iter()
        .map(CoreCompletionRequest::try_from)
        .collect()
}

/// Convert an upstream [`CoreBatchResult`] into the wire [`BatchResult`].
fn batch_result_from_core(core: CoreBatchResult) -> BatchResult {
    let responses = core
        .responses
        .into_iter()
        .map(|slot| match slot {
            Ok(resp) => BatchItem::Success {
                response: CompletionResponse::from(resp),
            },
            Err(err) => BatchItem::Failure {
                error_message: BlazenError::from(err).to_string(),
            },
        })
        .collect();
    BatchResult {
        responses,
        total_usage: core.total_usage.map(TokenUsage::from).unwrap_or_default(),
        total_cost_usd: core.total_cost.unwrap_or(0.0),
    }
}
